website/blog/2024/09-30-jsdom-based-scraping/index.md

```diff
@@ -0,0 +1,229 @@
+
+Then the most important part, we use `XMLHttpRequest.prototype.open` to extract the auth data and make call without actually using browsers or exposing the bot activity. 
+
+In the end of `createSessionFunction`, it returns a session with the required headers.
+
+Now coming to our main code, we will use CheerioCrawler and will use `prenavigationHooks` to inject the headers that we got from the earlier function into the `requestHandler`.
+
+```js
+const crawler = new CheerioCrawler({
+    sessionPoolOptions: {
+        maxPoolSize: 1,
+        createSessionFunction: async (sessionPool) => createSessionFunction(sessionPool, proxyConfiguration),
+    },
+    preNavigationHooks: [
+        (crawlingContext) => {
+            const { request, session } = crawlingContext;
+            request.headers = {
+                ...request.headers,
+                ...session.userData?.headers,
+
+            };
+        },
+    ],
+    proxyConfiguration,
+```
+
+Finally in the request handler we make the call using the headers and make sure how many calls are needed to fetch all the data handling pagination.
+
+```js
+async requestHandler(context) {
+        const { log, request, json } = context;
+        const { userData } = request;
+        const { itemsCounter = 0, resultsLimit = 0 } = userData;
+        if (!json.data) {
+            throw new Error('BLOCKED');
+        }
+        const { data } = json;
+        const items = data.list;
+        const counter = itemsCounter + items.length;
+        const dataItems = items.slice(0, resultsLimit && counter > resultsLimit ? resultsLimit - itemsCounter : undefined);
+        await Actor.pushData(dataItems);
+        const { pagination: { page, total } } = data;
+        log.info(`Scraped ${dataItems.length} results out of ${total} from search page ${page}`);
+        const isResultsLimitNotReached = counter < Math.min(total, resultsLimit);
+        if (isResultsLimitNotReached && data.pagination.has_more) {
+            const nextUrl = new URL(request.url);
+            nextUrl.searchParams.set('page', page + 1);
+            await crawler.addRequests([{
+                url: nextUrl.toString(),
+                headers: request.headers,
+                userData: {
+                    ...request.userData,
+                    itemsCounter: itemsCounter + dataItems.length,
+                },
+            }]);
+        }
+    },
+});
+```
+
+One important thing to note here is that we are making this code in a way that we can make any numbers of API calls. 
+
+In this particular example we just made one request and a single session, but you can make more if you need. When the first API call will be completed, it will create the second API call. Again, you can make more calls if needed, but we stopped at two.
+
+To make things more clear, here is how code flow looks:
+
+
```

Comment:
please pay some attention to this, sometimes you skip a blank line, sometimes you add two. its not a huge deal, rather a pet peeve :D 

```suggestion

```